commit 342b79cbab5335f7431530c4d9ee35f0bce97ede
Author: Phil Elwell <phil@raspberrypi.org>
Date:   Mon Jan 22 12:17:27 2018 +0000

    mmc: bcm2835-sdhost: Support underclocking
    
    Support underclocking of the SD bus in two ways:
    1. using the max-frequency DT property (which currently has no DT
       parameter), and
    2. using the exiting sd_overclock parameter.
    
    The two methods differ slightly - in the former the MMC subsystem is
    aware of the underclocking, while in the latter it isn't - but the
    end results should be the same.
    
    See: https://github.com/raspberrypi/linux/issues/2350
    
    Signed-off-by: Phil Elwell <phil@raspberrypi.org>

diff --git a/drivers/mmc/host/bcm2835-sdhost.c b/drivers/mmc/host/bcm2835-sdhost.c
index a9bc79bfdbb7..5b1aa59b84af 100644
--- a/drivers/mmc/host/bcm2835-sdhost.c
+++ b/drivers/mmc/host/bcm2835-sdhost.c
@@ -1515,8 +1515,7 @@ void bcm2835_sdhost_set_clock(struct bcm2835_host *host, unsigned int clock)
 	if (host->debug)
 		pr_info("%s: set_clock(%d)\n", mmc_hostname(host->mmc), clock);
 
-	if ((host->overclock_50 > 50) &&
-	    (clock == 50*MHZ))
+	if (host->overclock_50 && (clock == 50*MHZ))
 		clock = host->overclock_50 * MHZ + (MHZ - 1);
 
 	/* The SDCDIV register has 11 bits, and holds (div - 2).
@@ -1890,7 +1889,8 @@ int bcm2835_sdhost_add_host(struct bcm2835_host *host)
 
 	mmc = host->mmc;
 
-	mmc->f_max = host->max_clk;
+	if (!mmc->f_max || mmc->f_max > host->max_clk)
+		mmc->f_max = host->max_clk;
 	mmc->f_min = host->max_clk / SDCDIV_MAX_CDIV;
 
 	mmc->max_busy_timeout =  (~(unsigned int)0)/(mmc->f_max/1000);
